modin/pandas/test/utils.py

```diff
@@ -1228,23 +1228,35 @@ def teardown_test_files(test_paths: list):
         teardown_test_file(path)
 
 
-def sort_index_for_equal_values(series, ascending=False):
-    if series.index.dtype == np.float64:
+def sort_index_for_equal_values(df, ascending=True):
+    """Sort `df` indices of equal rows."""
+    if df.index.dtype == np.float64:
         # HACK: workaround for pandas bug:
         # https://github.com/pandas-dev/pandas/issues/34455
-        series.index = series.index.astype("str")
-    res = series.groupby(series, sort=False).apply(
+        df.index = df.index.astype("str")
+    res = df.groupby(by=df if df.ndim == 1 else df.columns, sort=False).apply(
         lambda df: df.sort_index(ascending=ascending)
     )
-    if res.index.nlevels > series.index.nlevels:
+    if res.index.nlevels > df.index.nlevels:
         # Sometimes GroupBy adds an extra level with 'by' to the result index.
         # GroupBy is very inconsistent about when it's doing this, so that's
         # why this clumsy if-statement is used.
         res.index = res.index.droplevel(0)
-    res.name = series.name
+    # GroupBy discards original index names, so restoring them here
+    if hasattr(df, "name"):
+        res.name = df.name
+    res.index.names = df.index.names
     return res
 
 
+def df_equals_with_non_stable_indices(df1, df2):
+    """Assert equality of two frames regardless of the index order for equal values."""
+    df1, df2 = map(try_cast_to_pandas, [df1, df2])
+    np.testing.assert_array_equal(df1.values, df2.values)
+    sorted1, sorted2 = map(sort_index_for_equal_values, [df1, df2])
+    df_equals(sorted1, sorted2)
```

Comment:
Why do we need to check equality of values twice? It looks expensive.

Reply:
1. The first equality check (ln. 1255) verifies that the actual frame's values are equals (omitting the indices). This step verifies that the Modin frame without any modification contains the valid data (but the actual order of indices may be different from pandas)
2. The second check (ln. 1257) verifies the equality of the modified results (sorted indices). This step mostly checks the validity of the index labels concordance with the frequency values.